File: frontend/src/app/core/mailbox.models.ts
/** A docker-mailserver mail account, as returned by /api/mailboxes. */
export interface Mailbox {
  email: string;
  domain: string;
  /** Dovecot quota (e.g. "5G"), or null when unlimited. */
  quota: string | null;
}

export interface MailboxCreateRequest {
  email: string;
  password: string;
  /** Optional quota (e.g. "5G"); omitted/empty means unlimited. */
  quota?: string | null;
}

export interface MailboxPasswordUpdateRequest {
  new_password: string;
}

export interface QuotaUpdateRequest {
  /** Quota (e.g. "5G"); null clears any existing limit. */
  quota: string | null;
}

/**
 * How much disk a mail account really occupies, as returned by
 * /api/mailboxes/usage. Distinct from `Mailbox.quota`, which is the configured
 * limit rather than the storage consumed.
 */
export interface MailboxUsage {
  email: string;
  used_bytes: number;
  /** The storage limit in bytes, or null when unlimited. */
  limit_bytes: number | null;
  /** Percentage of the limit consumed, or null when unlimited. */
  percent: number | null;
  message_count: number;
}

/** Disk usage of every mail account, ordered by usage, plus the totals. */
export interface MailboxUsageSummary {
  mailboxes: MailboxUsage[];
  total_used_bytes: number;
  /** Null as soon as one account is unlimited. */
  total_limit_bytes: number | null;
}

/** An alias address forwarding to a mailbox. */
export interface Alias {
  alias: string;
}

export interface AliasCreateRequest {
  alias: string;
}

/** The personal Sieve filter of one account (<email>.dovecot.sieve). */
export interface MailboxSieveScript {
  email: string;
  content: string;
  /** False when the account has no personal script yet. */
  configured: boolean;
  /** The script is compiled into the maildir when the mailserver starts. */
  restart_required: boolean;
}

export interface MailboxSieveScriptUpdateRequest {
  content: string;
}

File: backend/app/models/mailbox_models.py
"""API schemas for docker-mailserver mailbox management.

Mailboxes are not stored in the application database: they live in the
docker-mailserver flat files shared with the mailserver container through a
common volume:

* ``postfix-accounts.cf`` — one ``email|{scheme}hash`` line per account;
* ``postfix-virtual.cf``  — one ``alias target`` line per alias mapping;
* ``dovecot-quotas.cf``   — one ``email:quota`` line per quota-limited account.

These Pydantic schemas describe the request/response shapes only; persistence is
handled by :mod:`app.services.mailbox_service`.
"""

from typing import Annotated

from pydantic import BaseModel, EmailStr, Field

# A Dovecot quota, e.g. ``5G`` or ``512M``. Units: K/M/G/T (binary suffixes).
QuotaStr = Annotated[str, Field(pattern=r"^\d+[KMGT]$")]


class Mailbox(BaseModel):
    """A single mail account, as exposed by the API (never the password hash)."""

    email: EmailStr
    domain: str
    quota: str | None = None


class MailboxCreate(BaseModel):
    """Request schema for creating a mailbox."""

    email: EmailStr
    password: str = Field(min_length=8, max_length=1024)
    quota: QuotaStr | None = None


class MailboxPasswordUpdate(BaseModel):
    """Request schema for resetting a mailbox password."""

    new_password: str = Field(min_length=8, max_length=1024)


class QuotaUpdate(BaseModel):
    """Request schema for setting or clearing a mailbox quota.

    A ``null`` (or omitted) value clears any existing quota limit.
    """

    quota: QuotaStr | None = None


class MailboxUsage(BaseModel):
    """How much disk one mail account actually occupies, as Dovecot sees it.

    ``quota`` in :class:`Mailbox` is the *configured limit*; this is the storage
    really consumed, reported by ``doveadm quota get``.
    """

    email: str
    # Bytes currently stored in the maildir.
    used_bytes: int = 0
    # The account's storage limit in bytes, or ``None`` when unlimited.
    limit_bytes: int | None = None
    # Percentage of the limit consumed, or ``None`` when unlimited.
    percent: int | None = None
    # Messages currently stored in the maildir.
    message_count: int = 0


class MailboxUsageSummary(BaseModel):
    """Disk usage of every mail account plus the totals across all of them."""

    mailboxes: list[MailboxUsage] = Field(default_factory=list)
    total_used_bytes: int = 0
    # Sum of the configured limits; ``None`` as soon as one account is unlimited.
    total_limit_bytes: int | None = None


class Alias(BaseModel):
    """An alias address that forwards to a mailbox."""

    alias: EmailStr


class AliasCreate(BaseModel):
    """Request schema for adding an alias to a mailbox."""

    alias: EmailStr
